### PR TITLE
Add max code length tests, fix behavior for Python, Rust and C++

### PR DESCRIPTION
**Pull request by WilliamDenniss**
Replaces #287, moves tests to data files.

Includes #299 (I will rebase when that is merged).

### CODE REVIEW

**Comment by zongweil** [comment]
The C++ library seems to be failing the new tests. If you have time, would you mind looking into it? If not, I can take a look next week.

**Comment by WilliamDenniss** [comment]
Sure thing, I fixed the C++ issue (was returning the original size in the CodeArea object) by moving the truncation a bit earlier.

**Comment by WilliamDenniss** [comment]
@zongweil I believe this is good to go, all implementations are now passing the tests I added to verify the max code length behavior.

File: cpp/openlocationcode.cc [hunk 1]
```diff
@@ -199,7 +199,7 @@ std::string Encode(const LatLng &location) {
 
 CodeArea Decode(const std::string &code) {
   // Make a copy that doesn't have the separator and stops at the first padding
-  // character.
+  // character, and is constrained to the maximum length.
   std::string clean_code(code);
   clean_code.erase(
```
Comment (zongweil) [inline cpp/openlocationcode.cc]: The code from lines 203-210 seems to be duplicated in CodeLength(). Can you extract and de-dupe?

Comment (WilliamDenniss) [reply cpp/openlocationcode.cc]: Sure thing! Done.

**Comment by WilliamDenniss** [comment]
@zongweil this should be ready now, PTAL :)

File: cpp/openlocationcode.cc [hunk 2]
```diff
@@ -92,6 +92,19 @@ double adjust_latitude(double latitude_degrees, size_t code_length) {
   return latitude_degrees - precision / 2;
 }
 
+// Remove the separateor and padding characters from the code.
```
Comment (zongweil) [inline cpp/openlocationcode.cc]: separator

**Comment by zongweil** [comment]
Thanks! One typo, I'll merge it in right after :).

File: cpp/openlocationcode.cc (hunk 2 above)
Comment (WilliamDenniss) [reply cpp/openlocationcode.cc]: Done

**Pull request merged**